runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/WindmillStreamPoolSender.java

```diff
@@ -0,0 +1,234 @@
+      GetDataClient getDataClient,
+      HeartbeatSender heartbeatSender,
+      StreamingWorkScheduler streamingWorkScheduler,
+      Runnable waitForResources,
+      Function<String, Optional<ComputationState>> computationStateFetcher) {
+    this.isRunning = new AtomicBoolean(false);
+    this.hasGetWorkStreamStarted = new AtomicBoolean(false);
+    this.connection = connection;
+    this.getWorkRequest = getWorkRequest;
+    this.getWorkBudget = getWorkBudget;
+    this.streamingEngineStreamFactory = streamingEngineStreamFactory;
+    this.getDataClient = getDataClient;
+    this.heartbeatSender = heartbeatSender;
+    this.streamingWorkScheduler = streamingWorkScheduler;
+    this.waitForResources = waitForResources;
+    this.computationStateFetcher = computationStateFetcher;
+    this.workCommitter = workCommitter;
+    this.workProviderExecutor =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setPriority(Thread.MIN_PRIORITY)
+                .setNameFormat("DispatchThread")
+                .build());
+  }
+
+  public static WindmillStreamPoolSender create(
+      WindmillConnection connection,
+      GetWorkRequest getWorkRequest,
+      GetWorkBudget getWorkBudget,
+      GrpcWindmillStreamFactory streamingEngineStreamFactory,
+      WorkCommitter workCommitter,
+      GetDataClient getDataClient,
+      HeartbeatSender heartbeatSender,
+      StreamingWorkScheduler streamingWorkScheduler,
+      Runnable waitForResources,
+      Function<String, Optional<ComputationState>> computationStateFetcher) {
+    return new WindmillStreamPoolSender(
+        connection,
+        getWorkRequest,
+        new AtomicReference<>(getWorkBudget),
+        streamingEngineStreamFactory,
+        workCommitter,
+        getDataClient,
+        heartbeatSender,
+        streamingWorkScheduler,
+        waitForResources,
+        computationStateFetcher);
+  }
+
+  private void dispatchLoop() {
+    while (isRunning.get()) {
+      this.getWorkStream =
+          streamingEngineStreamFactory.createGetWorkStream(
+              connection.currentStub(), getWorkRequest, getWorkItemReceiver());
+      this.getWorkStream.start();
+      this.hasGetWorkStreamStarted.set(true);
+
+      try {
+        // Reconnect every now and again to enable better load balancing.
+        // If at any point the server closes the stream, we will reconnect immediately;
+        // otherwise
+        // we half-close the stream after some time and create a new one.
+        if (this.getWorkStream != null
+            && !this.getWorkStream.awaitTermination(
+                GET_WORK_STREAM_TIMEOUT_MINUTES, TimeUnit.MINUTES)) {
+          if (this.getWorkStream != null) {
+            this.getWorkStream.halfClose();
+          }
+        }
+      } catch (InterruptedException e) {
+        // Continue processing
+      }
+    }
+  }
+
+  private WorkItemReceiver getWorkItemReceiver() {
```

Comment:
I don't really like the duplication here between this an SingleSourceWorkerHarness in SE mode.
Could this just create a SingleSourceWorkerHarness and start it?

It seems like it is just missing setBudget support, but perhaps that could be added to SingleSourceWorkerHarness? Or alternatively could we have some static method to construct this receiver that we share between the two classes?